### PR TITLE
Button to display message source in the New Modmail

### LINKED ISSUE #323: Add "source" button to new modmail
Please add a `source` button to each modmail to view the Markdown source of the message.

Thanks!

P.S. I realize the `source` button that we all know and love is a RES feature, but [it doesn't look like RES is going to be supporting this feature on the redesign in general any time soon](https://github.com/honestbleeps/Reddit-Enhancement-Suite/issues/4951), let alone new modmail. And while I can almost entirely avoid the redesign as a moderator, I cannot avoid new modmail.

### PR DESCRIPTION
**Pull request by Xeoth**
If I did everything right, this fixes #323

### CODE REVIEW

File: extension/data/modules/newmodmailpro.js [hunk 1]
```diff
@@ -390,6 +390,40 @@ function newmodmailpro () {
...
+
+        // 'Source' button to display message source
+        window.addEventListener('TBNewPage', async event => {
+            if (event.detail.pageType === 'modmailConversation') {
+                // Add the button to the buttons over the message
+                $('.Thread__message').append('<button class="tb-source-button tb-action-button">Show Source</button>');
+
+                // Fetch and store the conversation info in cache
+                const currentID = event.detail.pageDetails.conversationID;
+                TBStorage.setCache('NewModmailPro', 'current-conversation', await TBApi.apiOauthGET(`/api/mod/conversations/${currentID}`).then(r => r.json()));
```
Comment (creesch) [inline extension/data/modules/newmodmailpro.js:402]: I am not sure we need to do the request for every modmail loaded. Possibly just do the request the first time source is clicked

File: extension/data/modules/newmodmailpro.js [hunk 2]
```diff
@@ -390,6 +390,40 @@ function newmodmailpro () {
                 }
             });
         }
+
+        // 'Source' button to display message source
+        window.addEventListener('TBNewPage', async event => {
+            if (event.detail.pageType === 'modmailConversation') {
+                // Add the button to the buttons over the message
+                $('.Thread__message').append('<button class="tb-source-button tb-action-button">Show Source</button>');
```
Comment (creesch) [inline extension/data/modules/newmodmailpro.js:398]: You also have conversations with collapsed messages for which you need to account. We already do this elsewhere in toolbox, I'll look it up in a bit.

**Comment by Xeoth** [comment]
Thanks for your comments, Creesch. I wouldn't have noticed those. I fixed what needed fixing and added a toggling feature. It should all be good now.

Please review and leave a comment if you find something that still needs a fix.

File: extension/data/modules/newmodmailpro.js [hunk 3]
```diff
@@ -390,6 +397,52 @@ function newmodmailpro () {
...
+
+                const $target = $(event.target);
+
+                $target.closest('.Thread__message').append('<button class="tb-source-button tb-action-button">Toggle Source</button>');
+
+                // Fetch and store the conversation info in cache
+                const currentID = event.detail.data.post.id;
+                TBStorage.setCache('NewModmailPro', 'current-conversation', await TBApi.apiOauthGET(`/api/mod/conversations/${currentID}`).then(r => r.json()));
+
+                $('.tb-source-button').unbind().click(async e => {
```
Comment (creesch) [inline extension/data/modules/newmodmailpro.js:415]: `unbind` is actually deprecated. You probably found it googling as the listener is triggered multiple times. A different solution would be to use `e.stopImmediatePropagation()`

File: extension/data/modules/newmodmailpro.js (hunk 1 above)
Comment (creesch) [reply extension/data/modules/newmodmailpro.js:413]: You haven't actually resolved this, for every message in a conversation you are still doing a request. 

My suggestion would be to set up your cache something along these lines: 

1. Keep a local boolean named `conversationCached` and set it to `false` by default.
2. Attach the buttons to the replies.
3. When a source button is clicked check the boolean, if true use cache for the source if false fetch the conversation and set the cache. 
4. Clear the boolean based on a `TBNewPage`  event.

File: extension/data/modules/newmodmailpro.js [hunk 4]
```diff
@@ -390,6 +397,52 @@ function newmodmailpro () {
...
+                    const $currentSourceBtn = $(e.currentTarget.parentElement);
+
+                    // Getting the ID of the message on which the button was clicked
+                    const [activeMessageID] = $currentSourceBtn.find('.Message__date')[0].pathname.split('/').slice(-1);
```
Comment (creesch) [inline extension/data/modules/newmodmailpro.js:419]: You already have the id from the jsAPI event, so this isn't really needed.

File: extension/data/modules/newmodmailpro.js [hunk 5]
```diff
@@ -390,6 +397,52 @@ function newmodmailpro () {
...
+
+        if (sourceButton) {
+            TB.listener.on('author', async event => {
+                if (event.detail.type !== 'TBmodmailCommentAuthor') {
+                    return;
+                }
+
+                const $target = $(event.target);
+
+                $target.closest('.Thread__message').append('<button class="tb-source-button tb-action-button">Toggle Source</button>');
```
Comment (creesch) [inline extension/data/modules/newmodmailpro.js:409]: I really don't like the placement of this button, my suggestion would be to add it to the message header and change the class to `tb-general-button`. 

Like so 

image

File: extension/data/styles/newmodmailpro.css [hunk 6]
```diff
@@ -36,6 +36,22 @@
     margin-bottom: 15px;
 }
 
+/* source textarea
+*/
+
+.tb-source-button {
```
Comment (creesch) [inline extension/data/styles/newmodmailpro.css:42]: Following my other comment this margin needs to change so it fits in the header.

The quote and Report button both have a left margin of 16px 

```css
body.mod-toolbox-rd .tb-source-button {
    margin-left: 16px;
}
```

**Review by creesch — CHANGES_REQUESTED** [review]
So uh yeah I ended up testing my suggested changes to make sure that what I was saying makes sense. 

So if you want you can simply use the below code ;) 

```javascript
        if (sourceButton) {
            let conversationCached = false;
            window.addEventListener('TBNewPage', () => {
                conversationCached = false;
            });

            TB.listener.on('author', event => {
                if (event.detail.type !== 'TBmodmailCommentAuthor') {
                    return;
                }
                const $target = $(event.target);

                $target.closest('.Message__header').append('<button class="tb-source-button tb-general-button">source</button>');

                $('.tb-source-button').click(async e => {
                    // Something is causing the listener to be triggered multiple times.
                    e.stopImmediatePropagation();
                    const $currentSourceBtn = $(e.currentTarget.parentElement);

                    // Getting the ID of the message on which the button was clicked.
                    const activeMessageID = event.detail.data.comment.id;
                    const $currentSourceField = $(`#tb-source-${activeMessageID}`);

                    // Toggling the source
                    if ($currentSourceField.length) {
                        // If the source field exists, toggle it

                        $currentSourceField.toggle();
                    } else {
                        // If the source field is not present (has not been requested yet), request it and create
                        // a div+textarea with the source.

                        if (!$currentSourceBtn.closest('.Thread__message').has('.tb-source-field').length) {
                            let conversationInfo;
                            if (conversationCached) {
                                conversationInfo = await TBStorage.getCache('NewModmailPro', 'current-conversation');
                            } else {
                                // Fetch and store the conversation info in cache
                                const currentID = event.detail.data.post.id;
                                conversationInfo = await TBApi.apiOauthGET(`/api/mod/conversations/${currentID}`).then(r => r.json());
                                TBStorage.setCache('NewModmailPro', 'current-conversation', conversationInfo);
                                conversationCached = true;
                            }
                            // Getting the body in markdown from selected message
                            const messageSource = conversationInfo.messages[activeMessageID].bodyMarkdown;

                            $currentSourceBtn.closest('.Thread__message').append(`
                                <div class="tb-source-field" id="tb-source-${activeMessageID}">
                                    <textarea readonly></textarea>
                                </div>`);
                            $(`#tb-source-${activeMessageID} textarea`).text(messageSource);
                        }
                    }
                });
            });
        }
```

**Review by creesch — APPROVED** [review]

**Pull request merged**